### PR TITLE
[Progression] Store testcase for regression testing using a signed url.

### PR DESCRIPTION
**Pull request by alhijazi**
The testcase is now stored under a randomly generated filename instead of the testcase file hash of contents.

### CODE REVIEW

**Comment by alhijazi** [comment]
@jonathanmetzman, I did not find a way to upload a file to gcs and to choose the filename at upload time using a signed url.
This, however, seems to be possible with the `storage.copy_file_to` by providing metadata as such:
`{'filename', 'some_generated_name_using_hash_of_contents'}`

**Comment by jonathanmetzman** [comment]
> @jonathanmetzman, I did not find a way to upload a file to gcs and to choose the filename at upload time using a signed url. This, however, seems to be possible with the `storage.copy_file_to` by providing metadata as such: `{'filename', 'some_generated_name_using_hash_of_contents'}`

I don't think this is possible with signed URLs.

**Comment by jonathanmetzman** [comment]
> > @jonathanmetzman, I did not find a way to upload a file to gcs and to choose the filename at upload time using a signed url. This, however, seems to be possible with the `storage.copy_file_to` by providing metadata as such: `{'filename', 'some_generated_name_using_hash_of_contents'}`
> 
> I don't think this is possible with signed URLs.

If it's really important we can maybe design an inteface for postprocess to rename the file.

File: src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py [hunk 1]
```diff
@@ -357,6 +359,29 @@ def _store_testcase_for_regression_testing(testcase, testcase_file_path):
     # Only store crashes with bugs associated with them.
     return
 
+  if not progression_task_input.HasField('regression_testcase_url'):
+    # No work to do.
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:363]: Remove this comment please. I don't think it's helpful.

File: src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py [hunk 2]
```diff
@@ -357,6 +359,29 @@ def _store_testcase_for_regression_testing(testcase, testcase_file_path):
...
+    return
+
+  regression_testcase_url = progression_task_input.regression_testcase_url
+
+  with open(testcase_file_path, 'rb') as testcase_file_handle:
+    testcase_file = testcase_file_handle.read()
+    if storage.upload_signed_url(testcase_file, regression_testcase_url):
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:370]: upload_signed_url will return True or exception, so it probably shouldn't return anything at all (if you want to be a hero you can change this in storage.py) so let's not depend on this behavior.

**Comment by alhijazi** [comment]
> > > @jonathanmetzman, I did not find a way to upload a file to gcs and to choose the filename at upload time using a signed url. This, however, seems to be possible with the `storage.copy_file_to` by providing metadata as such: `{'filename', 'some_generated_name_using_hash_of_contents'}`
> > 
> > 
> > I don't think this is possible with signed URLs.
> 
> If it's really important we can maybe design an inteface for postprocess to rename the file.

I think we can do just fine with randomly generated names here.

File: src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py [hunk 3]
```diff
@@ -357,6 +359,29 @@ def _store_testcase_for_regression_testing(testcase, testcase_file_path):
...
+
+  regression_testcase_url = progression_task_input.regression_testcase_url
+
+  with open(testcase_file_path, 'rb') as testcase_file_handle:
+    testcase_file = testcase_file_handle.read()
+    try:
+      storage.upload_signed_url(testcase_file, regression_testcase_url)
+      logs.log('Successfully stored testcase for regression testing: ' +
+               regression_testcase_url)
+    except Exception:
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:373]: We shouldn't catch `Exception` I probably wouldn't bother catching: https://google.github.io/styleguide/pyguide.html

Comment (alhijazi) [reply src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:373]: you mean just doing: 
```
try:
   ...
except:
    logs.log_error(...)
```

Comment (alhijazi) [reply src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:373]: The reason I am trying to catch is to keep the log_error.

File: src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py [hunk 4]
```diff
@@ -357,6 +359,29 @@ def _store_testcase_for_regression_testing(testcase, testcase_file_path):
...
+      logs.log_error('Failed to store testcase for regression testing: ' +
+                     regression_testcase_url)
+
+
+def _set_regression_testcase_upload_url(
+    progression_input: uworker_msg_pb2.ProgressionTaskInput,
+    testcase: data_types.Testcase):
+  """Determines and sets the signed regression_testcase_url (if any) in
+  the progression task input.
+  Raises RuntimeError in case of UUID collision on the generated filename.
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:383]: Kind of silly  that we check for this IMO

**Review by jonathanmetzman — APPROVED** [review]
LGTM

**Pull request merged**